ethereum_trie/binary.nim

```diff
@@ -1,158 +1,95 @@
 import
-  rlp, trie.constants, trie.validation, trie.exceptions, trie.utils.sha3,
-  trie.utils.binaries, trie.utils.nodes
+  keccak_tiny, types, utils/binaries, nodes, rlp/types as rlpTypes, utils
+
+export
+  types
 
 type
-  Hash = string
-
-  BinaryTrie* = object of object
-    db*: void
-    rootHash*: Hash
-
-proc makeBinaryTrie*(db: Any; rootHash: Hash): BinaryTrie =
-  result.db = db
-  result.rootHash = rootHash
-
-proc getImpl*(self: BinaryTrie; nodeHash: Hash; keypath: string): string =
-  ##         Note: keypath should be in binary array format, i.e., encoded by encode_to_bin()
-  if nodeHash == BLANKHASH:
-    return nil
-  (nodetype, leftChild, rightChild) = parseNode(self.db[nodeHash])
-  if nodetype == LEAFTYPE:
+  TrieNodeKey = object
+    hash: KeccakHash
+    usedBytes: uint8
+
+  BinaryTrie[DB: TrieDatabase] = object
+    dbLink: ref DB
+    rootHash: TrieNodeKey
+
+let
+  BLANK_HASH     = hashFromHex("c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470")
+  zeroBytesRange = Range[byte]()
+
+converter toTrieNodeKey(hash: KeccakHash): TrieNodeKey =
+  result.hash = hash
+  result.usedBytes = 32
+
+converter toTrieNodeKey(hash: BytesRange): TrieNodeKey =
+  assert hash.len == 32
+  for i in 0..<32:
+    result.hash.data[i] = uint8(hash[i])
+  result.usedBytes = 32
+
+proc initBinaryTrie*[DB](db: ref DB): BinaryTrie[DB] =
+  result.dbLink = db
+  result.rootHash = BLANK_HASH.toTrieNodeKey
+
+proc `==`(a, b: BytesRange): bool =
+  if a.len != b.len: return false
+  for i in 0..<a.len:
+    if a[i] != b[i]: return false
+  result = true
+
+proc getAux[DB](self: BinaryTrie[DB], nodeHash: TrieNodeKey, keyPath: BinVector): BytesRange =
```

Comment:
Prefer the implicit generic style when you don't need to reference the generic parameters:

``` nim
proc getAux(self: BinaryTrie, nodeHash: TrieNodeKey, keyPath: BinVector): BytesRange = 
```

This makes the code shorter and more resilient to refactoring (we can add an remove generic parameters without changing all the procs defined in the module).